Upgrade to OpenTracing 2.0 API

carlosalberto opened the pull request:
Creating the prototype of the OT 2.0 API update as a PR, in order to have a single place to discuss the proposed changes ;)

Latest iteration:
* `opentracing.tracer.scope_manager` is used for activation.
* `span_in_stack_context` requires a `TornadoScopeManager`.

@yurishkuro let me know ;)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber-common/opentracing-python-instrumentation?pullRequest=54) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber-common/opentracing-python-instrumentation?pullRequest=54) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber-common/opentracing-python-instrumentation?pullRequest=54) it.</sub>

carlosalberto commented:
Oh, didn't ran fine under Python 3.3, I wonder why, as it ran fine in all the other versions ;)

SEJeff commented:
@carlosalberto [Python 3.3.x is EOL](https://www.python.org/downloads/release/python-337/) as of Sept 2017, consider simply dropping it entirely.

carlosalberto commented:
Thanks for the info @SEJeff 

Maybe change that in Travis @yurishkuro ? ;)

yurishkuro commented:
@isaachier I seem to recall you having issues with py 3.3 elsewhere, was it similar to this?

```
Traceback:
<frozen importlib._bootstrap>:1519: in _find_and_load_unlocked
    ???
E   AttributeError: 'module' object has no attribute '__path__'
During handling of the above exception, another exception occurred:
tests/opentracing_instrumentation/test_install_client_hooks.py:25: in <module>
    from opentracing_instrumentation.client_hooks import install_client_interceptors
opentracing_instrumentation/__init__.py:21: in <module>
    from .request_context import get_current_span  # noqa
opentracing_instrumentation/request_context.py:26: in <module>
    from opentracing.ext.scope_manager.tornado import TornadoScopeManager
E   ImportError: No module named 'opentracing.ext.scope_manager.tornado'; opentracing.ext.scope_manager is not a package
```

yurishkuro commented:
Well, Java 1.6 is EOL for half a decade, and we still debate whether we should support it.  But I don't have objections to dropping 3.3

isaachier commented:
Well here the issue is worse because the tooling doesn't support anything anymore. Also, with Java the API changes are generally major/breaking changes. People generally only care if something runs on Python 2 or runs on Python 3, but it is rare to be specific about a minor version of Python in my experience.

wjoel commented:
Now that https://github.com/opentracing-contrib/python-flask has been upgraded to support OT 2.0 with the scope manager integration, would it be possible to pick this up again? I'd be happy to help as best I can.

Jamim commented on setup.py:
```diff
@@ -31,7 +31,7 @@
         'tornado>=4.1,<5',
         'contextlib2',
-        'opentracing>=1.1,<2',
+	'opentracing==2.0.0',
```
Hi @carlosalberto,
Please use spaces instead of `tab`.

Hello @yurishkuro,
I suppose that this `tab` came from [here](https://github.com/yurishkuro/opentracing-python-instrumentation/commit/8e22b09ff802860fa9d8e83f3df1409cc6b10b96#r31397576).

carlosalberto commented:
Hey @wjoel 

Oh, I think this instrumentation is for specifically using Jaeger + OT - for OpenTracing in general, the flask instrumentation you mentioned should be about fine.

@yurishkuro Is there something @wjoel or me could help with, to get this PR moving forward?

wjoel commented:
Hi @carlosalberto,

I don't see anything Jaeger specific in https://github.com/uber-common/opentracing-python-instrumentation/blob/master/setup.py#L33, so I thought it should work with any OT tracer. The readme doesn't mention Jaeger either, only OpenTracing. Seems like it should work?

Specifically, we'd like to use the instrumentation for `requests` and the decorators like https://github.com/uber-common/opentracing-python-instrumentation/blob/master/opentracing_instrumentation/local_span.py#L64

yurishkuro commented:
This instrumentation is intentionally independent of Jaeger. 

I had a 2.0-compatible fork of this PR working in https://github.com/yurishkuro/opentracing-python-instrumentation/tree/ot_scopemanager_integration. Unfortunately, we're currently short-handed and I can't dedicate time to finishing up the upgrade. If we get this PR green I am happy to merge. 

Side note: I have no idea why CLA is complaining about my commits, I don't see any in this PR (not to mention that I am part of Uber and signed it long time ago).

carlosalberto commented:
Hey @yurishkuro 

I've update this PR with your changes and made it work - still failing for Python 3.3 but that is EOL as @SEJeff mentioned already.

Good to go in?

yurishkuro commented:
Thanks, @carlosalberto , I'll review in the next couple days (my calendar is at 100% full today).

We will need the build to pass, please merge master where py3.3 is removed

carlosalberto commented:
Hey @yurishkuro thanks a lot, appreciated.

I've merged master and updated, and now Travis passes ;)

yurishkuro commented on requirements-test.txt:
```diff
@@ -3,3 +3,5 @@
 
 -e .[tests]
+
+-e git+https://github.com/carlosalberto/opentracing-python.git@scope_managers_integration#egg=opentracing
```
what is this for?

carlosalberto commented:
Hey @yurishkuro 

I updated the required items, and also updated `current_span()` so it checks for `RequestContextManager.current_context()` (the pre `ScopeManager` implementation), in case the users use `RequestContextManager` directly.

I felt like adding a deprecation note on using it directly, and ask users to use `span_in_context()`and `span_in_stack_context()` instead, as they are the ones that are exposed by default **and** they hide implementation details - but I don't know if that's something you would be ok with.

If you are ok with it, I will go ahead and update the README to use the mentioned functions instead of `RequestContextManager` (and perhaps deprecate it).

yurishkuro commented:
> I felt like adding a deprecation note on using it directly

+1

Jamim commented on tests/opentracing_instrumentation/test_traced_function_decorator.py:
```diff
@@ -30,6 +31,7 @@
 from opentracing_instrumentation import traced_function
 from opentracing_instrumentation import span_in_stack_context
 
+patch = mock.patch
```
I believe that this line isn't really required here since `patch` is used only once.
You can use `@mock.patch` instead.

carlosalberto commented:
Hey hey -

Have updated the documentation to reflect the latest changes (and have also mark as private a class in `request_context.py`, in case the users are still using it directly).

yurishkuro commented on opentracing_instrumentation/request_context.py:
```diff
@@ -23,11 +23,15 @@
 
 import opentracing
-import tornado.stack_context
+from opentracing.scope_managers.tornado import TornadoScopeManager
```
are we sure that TornadoScopeManager is thread-safe? Specifically, see #71

carlosalberto replied:
It should be, based on https://github.com/opentracing/opentracing-python/blob/master/opentracing/scope_managers/tornado.py#L154 - in any case, will double check by merging your latest test case and try things out ;)

yurishkuro commented:
Resolves #49

carlosalberto commented on opentracing_instrumentation/request_context.py:
```diff
@@ -210,8 +169,12 @@ def start_response_wrapper(status, response_headers,
+
+    # Return a no-op Scope if None was specified.
+    if span is None:
+        return opentracing.Scope(None, None)
```
It depends on the semantics of `ScopeManager.activate()` upon receiving `None` (some tracers may report this as an error), and thus I preferred to act defensively and return a no-op object.

The alternative would have been to report this as an error, but `span_in_context()` actually is expected to work with `span=None`.

carlosalberto commented on README.md:
```diff
@@ -38,7 +38,15 @@ Finally, a `@traced_function` decorator is provided for manual instrumentation.
 ### In-process Context Propagation
 
-`request_context` implements thread-local based context propagation facility.
+`span_in_context` implements context propagation using the current `opentracing.tracer.scope_manager`,
```
Yes, this is still about the `request_context` module. 

Sounds about right, will mention that and update the docs 👍

carlosalberto commented on README.md:
```diff
@@ -153,10 +165,7 @@ def run_coroutine_with_span(span, func, *args, **kwargs):
-    def mgr():
-        return RequestContextManager(span)
-
-    with tornado.stack_context.StackContext(mgr):
```
We are using internally now a subclass of `StackContext`, so contexts will be properly stashed/restored, but a few details will be different. Everything should work if the user uses the `span_in_stack_context()` function, though. But definitely needs clarification, just in case. Will add it 👍

carlosalberto commented:
Hey @yurishkuro I've merged master (along with your new threading test for Tornado based apps), and update the code/README to reflect the requested changes.

Let me know if you think we need further clarification or something of the sort ;)

yurishkuro commented:
@carlosalberto can you sign the CLA?

yurishkuro commented:
@carlosalberto ping

carlosalberto commented:
"You have signed the CLA for uber-common/opentracing-python-instrumentation" I get when I try to sign the CLA - Hey @yurishkuro Any idea? I will troubleshoot it tomorrow anyway.

yurishkuro commented:
Did you sign with the same email as what's registered with GitHub? Can you try to push another commit to rerun cla not?

carlosalberto commented:
Hey @yurishkuro I both pushed a new commit and verified I'm using the same email, and we are still stuck with the CLA :/ 

I hope this is not the same bug as https://github.com/cla-assistant/cla-assistant/issues/241

yurishkuro commented:
@carlosalberto turns out it was due to a couple of my commits!  Sorry about noise. I will merge this soon.
